### ISSUE TITLE
Question about the code in agent_encoder

### ISSUE THREAD

**Issue by Karim-SUN** [tsinghua-fib-lab/LCSim#7]
Hi, 
Firstly, I'd like to express my appreciation for your excellent LCSim project. It's valuable and quite useful for beginners in this area.

However, I came to some difficulties in figuring out how the model works like in `agent_encoder`. To be specific I found it hard to get the core idea when you encode the `relation for agents to historical steps` between line 174 and 199. Roughly speaking I could say it stacks the agents' position and heading information relative to historical time steps and then applies a Fourier_Embedding to get `r_t` tensor. But could you explain in a more detailed way how you process these data to get the final goal step by step?

```
        # relation for agents to historical steps
        pos_t = pos_a.reshape(-1, self.input_dim)
        head_t = head_a.reshape(-1)
        head_vector_t = head_vector_a.reshape(-1, 2)
        mask_t = mask.unsqueeze(2) & mask.unsqueeze(1)  # (A, T, T)
        mask_t[..., :-1] = False  # target only the last historical step
        edge_index_t = dense_to_sparse(mask_t)[0]
        edge_index_t = edge_index_t[:, edge_index_t[1] > edge_index_t[0]]
        edge_index_t = edge_index_t[
            :, edge_index_t[1] - edge_index_t[0] <= self.time_span
        ]
        rel_post_t = pos_t[edge_index_t[1]] - pos_t[edge_index_t[0]]
        rel_head_t = wrap_angle(head_t[edge_index_t[1]] - head_t[edge_index_t[0]])
        r_t = torch.stack(
            [
                torch.norm(rel_post_t[:, :2], p=2, dim=-1),
                angle_between_2d_vectors(
                    ctr_vector=head_vector_t[edge_index_t[1]],
                    nbr_vector=rel_post_t[:, :2],
                ),
                rel_head_t,
                edge_index_t[0] - edge_index_t[1],
            ],
            dim=-1,
        )
        r_t = self.r_t_emb(continuous_inputs=r_t, categorical_embs=None)
```

Thanks a lot  :)

**Comment by Karim-SUN** [comment]
So far from my understanding, it seems to mask out trajectories that are invalid until the present time step, and then gather all relative pos and heading infos under this mask.

**Comment by lionel-me** [comment]
Sorry for late response, and yes, your understanding is right, here we try to do the attention calculation only on the valid historical steps of each agent, and for each agent, the states are encoded in a relative form to its current coordinate. This part of scene encoder is inspired by the famous previous work [QCNet](https://github.com/ZikangZhou/QCNet), the more detailed design and why we do that can be found in this work.

**Comment by 2150016-njl** [comment]
您好！
您在论文中提出的基于图的编解码器让我学习到很多知识，也是首次将扩散模型 EDM 架构用于交通模拟领域的工作，让我受益匪浅。

但是，我在复现模型过程中，始终会出现低速行驶的场景朝向摆动剧烈的现象，因为代码中的朝向角是通过运动向量计算得到的，因此我认为是轨迹预测准确性不足的原因，在低速行驶中会预测出折线轨迹，导致朝向摆动剧烈。
最初，我认为是数据不足的原因，导致模型没有充分训练，我不断地增加数据，目前已经使用了waymo motion dataset v1.2的1000个scenario类型文件，三个 target 都存在比较严重的问题。我和您项目中不同的配置只有显卡数量，我只有 1 张 4090 显卡，为了节省时间，我只训练不推理，训练一个 target 的模型大概需要将近三周。我在训练扩散模型前预训练了 PCA 权重。

我想向您请教一下出现这个问题可能的原因。如果方便的话，您能否提供模型权重以供参考学习呢？
十分感谢！！！
